src/main/java/com/aws/greengrass/clientdevices/auth/PermissionEvaluationUtils.java

```diff
@@ -139,14 +150,58 @@ private static Resource parseResource(String resourceStr) {
 
         Matcher matcher = SERVICE_RESOURCE_PATTERN.matcher(resourceStr);
         if (matcher.matches()) {
-            return Resource.builder().service(matcher.group(1)).resourceType(matcher.group(2))
-                    .resourceName(matcher.group(3)).build();
+            return Resource.builder().service(matcher.group(1))
+                    .resourceType(matcher.group(2))
+                    .resourceName(matcher.group(3))
+                    .build();
         }
 
         throw new IllegalArgumentException(
                 String.format("Resource %s is not in the form of %s", resourceStr, SERVICE_RESOURCE_PATTERN.pattern()));
     }
 
+    /**
+     * utility method to permission with variable value.
+     *
+     * @param session    current device session
+     * @param permission permission to parse
+     * @return updated permission
+     */
+
+    static Permission updateResource(@NonNull Session session, @NonNull Permission permission) {
+
+        String resource = permission.getResource();
+
+        Pattern pattern = Pattern.compile(POLICY_VARIABLE_PATTERN);
```

Comment:
Store the compiled pattern as a constant field. Compiling the pattern every time this method is called is not necessary.